navbar: Update left and right sidebar icons with dot and solid style.

ZetaSpark opened the pull request:
This PR introduces dynamic toggling functionality for the left and right sidebar icons, improving user interactivity based on the sidebar's state. The previous Font Awesome icon have been replaced with modern SVG icons for better visual quality and enhanced performance

Fixes: #32825 
**Screenshots and screen captures:**

| Icon Type      | Open State Icon             | Closed State Icon          |
|----------------|-----------------------------|----------------------------|
| Left Sidebar   | ![Screenshot 2025-01-07 201905](https://github.com/user-attachments/assets/14946925-b88f-4402-8b4e-7f6f3b9ea286)  | ![Screenshot 2025-01-07 201915](https://github.com/user-attachments/assets/7d5e6d12-8972-4f80-9413-eb3a6d3d723b)  |
| Right Sidebar  | ![Screenshot 2025-01-07 201936](https://github.com/user-attachments/assets/5373cab8-674f-4d1c-a1ec-1e5d8a5e607c) | ![Screenshot 2025-01-07 201946](https://github.com/user-attachments/assets/974dd629-324c-47a8-979c-673c7d3ccfa8)  |





<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

ZetaSpark commented:
Hi @amanagr pls review my PR.
I tried to follow a similar approach to the existing codebase for handling .svg files, and the toggle functionality is now managed using the hidden class that was already defined in the codebase.

amanagr commented on web/templates/navbar.hbs:
```diff
@@ -13,7 +14,9 @@
                 <div id="streamlist-toggle" class="tippy-zulip-delayed-tooltip {{#if embedded}}hide-streamlist-toggle-visibility{{/if}}" data-tooltip-template-id="show-left-sidebar-tooltip-template">
-                    <a class="left-sidebar-toggle-button" role="button" tabindex="0"><i class="fa fa-reorder" aria-hidden="true"></i>
+                    <a class="left-sidebar-toggle-button" role="button" tabindex="0">
+                        <i class="zulip-icon zulip-icon-panel-left" id="zulip-icon-panel-left" style="display: none;"></i>
```
le's increase there font size to 18px and not use `style` attribute here but instead use CSS.

ZetaSpark commented:
Hi @amanagr 

I’ve updated the font size to 18px by adding the necessary CSS classes in the relevant files.

While testing locally, I noticed that upon refreshing, the left and right sidebars retain their previously stored states, but the toggle icons were not in sync with them. I’ve addressed this issue as well and ensured the toggle icons now correctly reflect the sidebar states.

ZetaSpark commented on web/src/sidebar_ui.ts:
```diff
@@ -128,45 +161,52 @@ export function initialize(): void {
         e.stopPropagation();
 
+        const isHidden = $("body").hasClass("hide-right-sidebar");
         if (window.innerWidth >= media_breakpoints_num.xl) {
```
I’d like your guidance on handling `innerWidth`. Specifically, how should the toggle icons behave on different screen widths? The current code functions well on a normal viewport.

Thanks!

amanagr replied:
Well, you just need to come up with logic to restore the original behaviour.

amanagr commented on web/templates/navbar.hbs:
```diff
@@ -2,7 +2,8 @@
         <div class="column-left">
             <a class="left-sidebar-toggle-button {{#if embedded}}hide-streamlist-toggle-visibility{{/if}}" tabindex="0" role="button">
-                <i class="fa fa-reorder" aria-hidden="true"></i>
+                <i class="zulip-icon zulip-icon-panel-left" id="zulip-icon-panel-left" ></i>
+                <i class="zulip-icon zulip-icon-panel-left-dashed" id="zulip-icon-panel-left-dashed" ></i>
```
We don't need two elements here, we can just change the icon being displayed.

ZetaSpark commented:
Hi @amanagr
I have revised the code and now using only 1 `element` for toggling.
This is code also ensures proper `syncing` b/w states of icon and sidebars, `restoring` states of icon and sidebar states on re-rendering and have taken care the `responsiveness` as well.

Thanks!

ZetaSpark commented:
Hi @alya based on your suggestions I have added a commit to solve unread count toggle functionality in narrow screens.
I have also uploaded video [CZO](https://chat.zulip.org/#narrow/channel/101-design/topic/groups.20icon.3F) to show how it is currently working in zulip and how it should.

Thanks!

amanagr commented on web/src/sidebar_ui.ts:
```diff
@@ -31,19 +31,27 @@ function save_sidebar_toggle_status(): void {
-}
 
+    // Update the icon state after restoring sidebar states
+    updateLeftSidebarIcon();
```
We use snake case in Zulip.

ZetaSpark replied:
Thanks for pointing out, I will correct this.

amanagr commented:
I was wonder what a  CSS based solution for this would look like that just uses the existing JS? (even if we had to use two eleemnts for this)

ZetaSpark commented:
Hi @amanagr 
Initially, I also went with that `approach` and uses both icons as elements. But when I `dig` further into codebase, I found it's not that simple.
Apart from toggling we need to ensure:
1. `restoring` of states of icons when a user `refreshes` the page or open it later. (It should match with sidebar state as well)
2. `responsiveness` as well for different `viewports`, here icons just need to response in a way to match sidebar's response.
3. Unread count icon should also work properly for all viewports.

Currently we use `hide-right/left-sidebar` class to toggle sidebars in normal view and `expanded` class to toggle sidebars in narrow windows (Like add/remove these classes to hide/show sidebars in that viewport). So, I created 2 fns which show/hide icons based on which viewport we are currently on and which class is currently there (hide-right/left-sidebar or expanded) and since we already storing states of sidebars in local storage, we just need to call these fns there to restore the icon states.

So, I find using `JS` to be more efficient to define fns, handling every situation properly to `dynamically` toggle icons.

Pls, guide any other approach based on `CSS` which we can use, I would be happy to learn.

Thanks for your time.

amanagr commented:
Can you adjust your code to use something like this? (only did the part for left sidebar)

```diff
diff --git a/web/styles/zulip.css b/web/styles/zulip.css
index 7c59ac93e7..1ed7801aed 100644
--- a/web/styles/zulip.css
+++ b/web/styles/zulip.css
@@ -1629,7 +1629,23 @@ div.toggle_resolve_topic_spinner .loading_indicator_spinner {
     visibility: hidden;
 }
 
+.left-sidebar-toggle-button .zulip-icon-panel-left-dashed {
+    display: none;
+}
+
+.left-sidebar-toggle-button .zulip-icon-panel-left {
+    display: block;
+}
+
 %hide-left-sidebar {
+    .left-sidebar-toggle-button .zulip-icon-panel-left-dashed {
+        display: block;
+    }
+
+    .left-sidebar-toggle-button .zulip-icon-panel-left {
+        display: none;
+    }
+
     .app-main .column-left {
         display: none;
 
@@ -2093,6 +2109,8 @@ body:not(.hide-left-sidebar) {
         font-size: 18px;
     }
 
+    .zulip-icon-panel-left,
+    .zulip-icon-panel-left-dashed,
     .zulip-icon-help-bigger,
     .zulip-icon-user-list {
         font-size: 20px;
diff --git a/web/templates/navbar.hbs b/web/templates/navbar.hbs
index 9c965fb812..477f7b3490 100644
--- a/web/templates/navbar.hbs
+++ b/web/templates/navbar.hbs
@@ -1,8 +1,9 @@
 <div class="header">
     <nav class="header-main" id="top_navbar">
         <div class="column-left">
-            <a class="left-sidebar-toggle-button {{#if embedded}}hide-streamlist-toggle-visibility{{/if}}" tabindex="0" role="button">
-                <i class="fa fa-reorder" aria-hidden="true"></i>
+            <a class="header-button left-sidebar-toggle-button {{#if embedded}}hide-streamlist-toggle-visibility{{/if}}" tabindex="0" role="button">
+                <i class="zulip-icon zulip-icon-panel-left-dashed"></i>
+                <i class="zulip-icon zulip-icon-panel-left"></i>
                 <span class="left-sidebar-toggle-unreadcount">0</span>
             </a>
             <a href="" class="brand no-style">
@@ -63,7 +64,8 @@
             </div>
             <div id="userlist-toggle" class="hidden-for-spectators">
                 <a id="userlist-toggle-button" role="button" class="header-button navbar-item" tabindex="0">
-                    <i class="zulip-icon zulip-icon-user-list"></i>
+                    <i class="zulip-icon zulip-icon-panel-right-dashed"></i>
+                    <i class="zulip-icon zulip-icon-panel-right"></i>
                 </a>
             </div>
         </div>

```

amanagr commented on web/templates/navbar.hbs:
```diff
@@ -41,11 +42,6 @@
                     {{t 'Log in' }}
                 </a>
             </div>
-            <div id="userlist-toggle" class="hidden-for-spectators">
-                <a id="userlist-toggle-button" role="button" class="header-button navbar-item" tabindex="0">
-                    <i class="zulip-icon zulip-icon-user-list"></i>
-                </a>
-            </div>
```
Let's make the moving of the icon to be the right most be a separate commit.

ZetaSpark commented:
Hi @amanagr,
I am unable to get desired results by implementing your method in the right sidebar.

amanagr commented:
Thanks @ZetaSpark for your effort on this. Closing in favour a CSS based solution in #33235
